modules/dnn/src/layers/scatter_layer.cpp

```diff
@@ -0,0 +1,218 @@
+        for (size_t i = 0; i < inputs[0].size(); i++)
+        {
+            CV_CheckGE(inputs[0][i], inputs[1][i], "Scatter: each dim of input data should be greater than (or equal to) indices'.");
+            CV_CheckEQ(inputs[1][i], inputs[2][i], "Scatter: each dim of indices should be equal to updates'.");
+        }
+        outputs.assign(1, inputs[0]);
+        return false;
+    }
+
+    virtual void finalize(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr) CV_OVERRIDE
+    {
+        std::vector<Mat> inputs;
+        inputs_arr.getMatVector(inputs);
+
+        int ndims = inputs[0].dims;
+        current_indices.resize(ndims);
+    }
+
+    void forward(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr, OutputArrayOfArrays internals_arr) CV_OVERRIDE
+    {
+        CV_TRACE_FUNCTION();
+        CV_TRACE_ARG_VALUE(name, "name", name.c_str());
+
+        std::vector<Mat> inputs, outputs;
+        inputs_arr.getMatVector(inputs);
+        outputs_arr.getMatVector(outputs);
+
+        const Mat& data = inputs[0];
+        const Mat& indices = inputs[1];
+        const Mat& updates = inputs[2];
+        Mat& out = outputs[0];
+
+        typeDispatch(outputs[0].type(), data, indices, updates, out);
+    }
+
+    template<typename T, typename Functor>
+    void forward_impl(const Functor& rd, const Mat& data, const Mat& indices, const Mat& updates, Mat& out)
+    {
+        data.copyTo(out);
+
+        const int ndims = data.dims;
+        const int* shape = data.size.p;
+        const size_t* step = data.step.p;
+
+        const int* ind_shape = indices.size.p;
+        const size_t* ind_step = indices.step.p;
+
+        size_t inp_offset = 0;
+        size_t ind_offset = 0;
+        const T* p_index = indices.ptr<const T>();
+        const T* p_update = updates.ptr<const T>();
+        T* p_out = out.ptr<T>();
+
+        size_t total = indices.total();
+
+        int j, offset_at_idx, index;
+        size_t t, idx;
+        for (size_t i = 0; i < total; i++)
+        {
+            t = i;
+            for (j = ndims - 1; j >= 0; j--)
+            {
+                idx = t / ind_shape[j];
+                offset_at_idx = (int)(t - idx * ind_shape[j]);
+                current_indices[j] = offset_at_idx;
+                t = idx;
+            }
+
+            ind_offset = 0;
+            for (j = 0; j < ndims; j++)
+                ind_offset += current_indices[j] * ind_step[j];
+            ind_offset /= sizeof(T);
+
+            // get index and overwrite current indices
+            const T* tmp_p_index = p_index + ind_offset;
+            index = (int)(*tmp_p_index);
+            CV_Assert(index < shape[axis] && index > -shape[axis]);
+            current_indices[axis] = (index + shape[axis]) % shape[axis];
+
+            inp_offset = 0;
+            for (j = 0; j < ndims; j++)
+                inp_offset += current_indices[j] * step[j];
+            inp_offset /= sizeof(T);
```

Comment:
Seems like here we can also get rid of `current_indices` and `finalize()` and also get 2x speedup:

```cpp
t = i;
int at_axis_val = 0;
ind_offset = 0;
inp_offset = 0;
for (j = ndims - 1; j >= 0; j--)
{
    idx = t / ind_shape[j];
    offset_at_idx = (int)(t - idx * ind_shape[j]);
    ind_offset += offset_at_idx * ind_step[j];
    inp_offset += offset_at_idx * step[j];
    t = idx;
    if (j == axis)
    {
        at_axis_val = offset_at_idx * step[j];
    }
}

ind_offset /= sizeof(T);

// get index and overwrite current indices
const T* tmp_p_index = p_index + ind_offset;
index = (int)(*tmp_p_index);
CV_Assert(index < shape[axis] && index > -shape[axis]);
inp_offset = inp_offset - at_axis_val + ((index + shape[axis]) % shape[axis]) * step[axis];
inp_offset /= sizeof(T);
```

Reply:
Yes, another speedup of ~1.5X.